docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,404 @@
+
+- The DSCP marking does not need to be handled or acted upon by OVN-Kubernetes, just added to selected headers.
+
+## Introduction
+
+To enable NetworkQoS, we will use Differentiated Services Code Point (DSCP) which allows us to classify packets by setting a 6-bit field in the IP header, effectively marking the priority of a given packet relative to other packets as "Critical", "High Priority", "Best Effort" and so on.
+
+## User-Stories/Use-Cases
+#### Story 1
+
+As a user of OVN-Kubernetes, I should be able to mark egress traffic coming from a specific namespace with a valid DSCP value.
+
+#### Story 2
+
+As a user of OVN-Kubernetes, I should be able to apply bandwidth limit (rate and burst) for a pod traffic originating from a specific namespace.
+
+#### Story 3
+
+As a user of OVN-Kubernetes, I should be able to create multiple NetworkQoS objects per namesapce containing rules with different priorities.
+
+#### Story 4
+
+The NetworkQoS Status field must reflect observed state of the object covering entire Kubernetes cluster.
+
+## Proposed Solution
+
+By introducing a new CRD `NetworkQoS`, users could specify a DSCP value for packets originating from pods on a given namespace heading to a specified CIDR, Protocol and Port. This
+also supports metering for the packets by specifying bandwidth parameters `rate` and/or `burst`.
+The CRD will be Namespaced, with multiple resources allowed per namespace.
+The resources will be watched by ovn-k, which in turn will configure OVN's [QoS Table](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#NetworkQoS_TABLE).
+The `NetworkQoS` also has `status` field which is populated by ovn-k which helps users to identify whether NetworkQoS rules are configured correctly in OVN or not.
+
+ovn-controller has this configuration, ovn-encap-tos:
+
+-------8<-------------8<--------
+external_ids:ovn-encap-tos
+ovn-encap-tos indicates the value to be applied to OVN
+tunnel interface’s option:tos as specified in the
+Open_vSwitch database Interface table. Please refer to
+Open VSwitch Manual for details.
+
+   options : tos: optional string
+          Optional. The value of the ToS bits to be set on the encapsulat‐
+          ing packet. ToS is interpreted as DSCP and ECN  bits,  ECN  part
+          must be zero. It may also be the word inherit, in which case the
+          ToS will be copied from the inner packet if it is IPv4  or  IPv6
+          (otherwise  it  will be 0). The ECN fields are always inherited.
+          Default is 0.
+-------8<-------------8<--------
+
+which when set to inherit will take the marking from the inner header and apply it to the GENEVE packet. Therefore, east west traffic within the cluster can benefit
+from this API and not just the north south traffic.
+
+### API Details
+
+* A new API `NetworkQoS` under the `k8s.ovn.org/v1` version will be added to `go-controller/pkg/crd/networkqos/v1`.
```

Comment:
should we jump to `v1`, directly?

Should this be v1beta1 or some such?

Reply:
I see other CRDs are also just started with `v1` directly, If this is not good, please let me know, would change it to `v1beta1`